File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/python/ondc_signature.py
"""
ONDC Ed25519 Signature Authentication Module - Python Reference Implementation

Implements the ONDC protocol signature scheme for V3 API authentication.
"""

import base64
import hashlib
import json
from datetime import datetime, timezone
from typing import Dict, Optional, Tuple, Union
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives import serialization


class ONDCSignature:
    """ONDC Signature Generator for V3 API Authentication"""
    
    def __init__(self, subscriber_id: str, unique_key_id: str, private_key_seed: bytes):
        """
        Initialize ONDC signature generator.
        
        Args:
            subscriber_id: Unique subscriber ID
            unique_key_id: Unique key identifier
            private_key_seed: 32-byte seed for Ed25519 key generation
        """
        self.subscriber_id = subscriber_id
        self.unique_key_id = unique_key_id
        
        if not isinstance(private_key_seed, bytes) or len(private_key_seed) != 32:
            raise ValueError(f"private_key_seed must be exactly 32 bytes, got {len(private_key_seed)}")
        
        self.private_key = Ed25519PrivateKey.from_private_bytes(private_key_seed)
        self.public_key = self.private_key.public_key()
        self.public_key_b64 = self._get_public_key_b64()
    
    def _get_public_key_b64(self) -> str:
        """Get DER/SPKI-encoded public key (base64)."""
        public_bytes = self.public_key.public_bytes(
            encoding=serialization.Encoding.DER,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        return base64.b64encode(public_bytes).decode('utf-8')
    
    def _create_digest(self, body: Optional[Union[Dict, str, bytes]] = None) -> str:
        """Create BLAKE2b-512 digest of request body."""
        if body is None or body == {} or body == b'' or body == '':
            raw = b''
        elif isinstance(body, bytes):
            raw = body
        elif isinstance(body, str):
            raw = body.encode('utf-8')
        else:
            raw = json.dumps(body, separators=(', ', ': '), sort_keys=True, ensure_ascii=False).encode('utf-8')

        digest = hashlib.blake2b(raw, digest_size=64).digest()
        return base64.b64encode(digest).decode('utf-8')
    
    def _create_signing_string(self, created: int, expires: int, digest: str) -> str:
        """Create the signing string according to ONDC spec."""
        return f"(created): {created}\n(expires): {expires}\ndigest: BLAKE-512={digest}"
    
    def _sign_string(self, signing_string: str) -> str:
        """Sign the signing string with Ed25519 private key."""
        signature = self.private_key.sign(signing_string.encode('utf-8'))
        return base64.b64encode(signature).decode('utf-8')
    
    def generate_signature_header(
        self,
        body: Optional[Union[Dict, str, bytes]] = None,
        created: Optional[int] = None,
        expires: Optional[int] = None
    ) -> Tuple[str, str, int, int]:
        """
        Generate complete Signature authorization header and Digest header.
        
        Returns:
            Tuple of (authorization_header, digest_header, created_timestamp, expires_timestamp)
        """
        # Generate timestamps
        if created is None:
            now = datetime.now(timezone.utc)
            created = int(now.timestamp())
        if expires is None:
            # 60-second validity window (ONDC recommended: 30-60 seconds)
            expires = created + 60
        
        # Create digest
        digest = self._create_digest(body)
        digest_header = f"BLAKE-512={digest}"
        
        # Create signing string
        signing_string = self._create_signing_string(created, expires, digest)
        
        # Generate signature
        signature = self._sign_string(signing_string)
        
        # Construct Signature header
        key_id = f"{self.subscriber_id}|{self.unique_key_id}|ed25519"
        
        header = (
            f'Signature keyId="{key_id}",'
            f'algorithm="ed25519",'
            f'created="{created}",'
            f'expires="{expires}",'
            f'headers="(created) (expires) digest",'
            f'signature="{signature}"'
        )
        
        return header, digest_header, created, expires

File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/nodejs/ondc-signature.js
/**
 * ONDC Ed25519 Signature Generator - Node.js Implementation
 * 
 * Implements the ONDC protocol signature scheme for V3 API authentication.
 */

import * as ed25519 from '@noble/ed25519';
import blake2b from 'blake2b';
import { createHash } from 'crypto';

// Set SHA-512 for ed25519 (required by @noble/ed25519 v2+)
ed25519.etc.sha512Sync = (...messages) => {
    const hash = createHash('sha512');
    for (const message of messages) hash.update(message);
    return Uint8Array.from(hash.digest());
};

class ONDCSignature {
    /**
     * Initialize ONDC signature generator.
     *
     * @param {string} subscriberId - Unique subscriber ID
     * @param {string} uniqueKeyId - Unique key identifier
     * @param {Buffer} privateKeySeed - 32-byte private key seed
     */
    constructor(subscriberId, uniqueKeyId, privateKeySeed) {
        this.subscriberId = subscriberId;
        this.uniqueKeyId = uniqueKeyId;
        
        if (!Buffer.isBuffer(privateKeySeed) || privateKeySeed.length !== 32) {
            throw new Error(`privateKeySeed must be exactly 32 bytes, got ${privateKeySeed.length}`);
        }
        
        this.privateKey = privateKeySeed;
        this.publicKey = ed25519.getPublicKey(this.privateKey);
        this.publicKeyB64 = this._getPublicKeyB64();
    }
    
    /**
     * Get DER/SPKI-encoded public key (base64).
     */
    _getPublicKeyB64() {
        // SPKI/DER header for Ed25519 public key
        const spkiHeader = Buffer.from([
            0x30, 0x2a,  // SEQUENCE, 42 bytes
            0x30, 0x05,  // SEQUENCE, 5 bytes
            0x06, 0x03, 0x2b, 0x65, 0x70,  // OID 1.3.101.112 (Ed25519)
            0x03, 0x21, 0x00  // BIT STRING, 33 bytes (including 0x00 padding)
        ]);
        
        const spkiEncoded = Buffer.concat([spkiHeader, Buffer.from(this.publicKey)]);
        return spkiEncoded.toString('base64');
    }
    
    /**
     * Create BLAKE2b-512 digest of request body.
     */
    _createDigest(body) {
        let bodyBytes;
        
        if (body === null || body === undefined || body === '') {
            bodyBytes = Buffer.from('');
        } else if (Buffer.isBuffer(body)) {
            bodyBytes = body;
        } else if (typeof body === 'string') {
            bodyBytes = Buffer.from(body, 'utf8');
        } else if (typeof body === 'object') {
            // Sort object keys recursively for consistency
            const sortedBody = this._sortObjectKeys(body);
            // Serialize with specific formatting (matches Python's json.dumps with sort_keys=True)
            // Use precise regex to only replace JSON structure separators, not colons/commas in values
            const json = JSON.stringify(sortedBody, null, 0)
                .replace(/":/g, '": ')  // Replace key-value separator only
                .replace(/,"/g, ', "')  // Replace element separator before string
                .replace(/},{/g, '}, {'); // Replace array element separator
            bodyBytes = Buffer.from(json, 'utf8');
        } else {
            throw new Error('Body must be null, string, Buffer, or object');
        }
        
        // Generate BLAKE2b-512 hash
        const hash = blake2b(64).update(bodyBytes).digest();
        return Buffer.from(hash).toString('base64');
    }
    
    /**
     * Create the signing string according to ONDC spec.
     */
    _createSigningString(created, expires, digest) {
        return `(created): ${created}\n(expires): ${expires}\ndigest: BLAKE-512=${digest}`;
    }
    
    /**
     * Sign the signing string with Ed25519 private key.
     */
    async _signString(signingString) {
        const message = Buffer.from(signingString, 'utf8');
        const signature = await ed25519.sign(message, this.privateKey);
        return Buffer.from(signature).toString('base64');
    }
    
    /**
     * Generate complete Signature authorization header.
     *
     * @param {*} body - Request body (can be null, string, Buffer, or object)
     * @param {number} created - Unix timestamp when signature was created (null for current time)
     * @param {number} expires - Unix timestamp when signature expires (null for created + 300)
     * @returns {Promise<string>} Authorization header string
     */
    async generateSignatureHeader(body, created = null, expires = null) {
        // Generate timestamps
        if (created === null) {
            created = Math.floor(Date.now() / 1000);
        }
        if (expires === null) {
            // 60-second validity window (ONDC recommended: 30-60 seconds)
            expires = created + 60;
        }
        
        // Create digest
        const digest = this._createDigest(body);
        
        // Create signing string
        const signingString = this._createSigningString(created, expires, digest);
        
        // Generate signature
        const signature = await this._signString(signingString);
        
        // Construct Signature header
        const keyId = `${this.subscriberId}|${this.uniqueKeyId}|ed25519`;
        
        return `Signature keyId="${keyId}",algorithm="ed25519",created="${created}",expires="${expires}",headers="(created) (expires) digest",signature="${signature}"`;
    }
    
    /**
     * Generate Digest header.
     */
    generateDigestHeader(body) {
        return `BLAKE-512=${this._createDigest(body)}`;
    }
    
    /**
     * Get public key (base64).
     */
    getPublicKey() {
        return this.publicKeyB64;
    }
    
    /**
     * Serialize body to JSON string matching the format used in digest calculation.
     * Use this when sending HTTP requests to ensure digest matches.
     * 
     * @param {*} body - Request body
     * @returns {string} Serialized JSON string
     */
    serializeBody(body) {
        if (body === null || body === undefined || body === '') {
            return '';
        } else if (typeof body === 'string') {
            return body;
        } else if (typeof body === 'object') {
            const sortedBody = this._sortObjectKeys(body);
            const json = JSON.stringify(sortedBody, null, 0)
                .replace(/":/g, '": ')  // Replace key-value separator only
                .replace(/,"/g, ', "')  // Replace element separator before string
                .replace(/},{/g, '}, {'); // Replace array element separator
            return json;
        }
        throw new Error('Body must be null, string, or object');
    }
    
    /**
     * Helper: Recursively sort object keys alphabetically for consistency.
     */
    _sortObjectKeys(obj) {
        if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
            return obj;
        }
        
        const sorted = {};
        Object.keys(obj).sort().forEach(key => {
            sorted[key] = this._sortObjectKeys(obj[key]);
        });
        return sorted;
    }
}

export default ONDCSignature;
